### DISCUSSION TITLE
Problems getting non-interactive authorization to token to work.

### DISCUSSION THREAD

**Discussion by bbakkebo** [zitadel/zitadel#4814] [Q&A]
Hi,

I have use case where we need to login a user and get a token back without having to login through web or user agents.
So i have tried a few variations of this below:
![image](https://user-images.githubusercontent.com/77129/205275175-9db1f397-86e4-4f9f-994a-f2e065258e9e.png)

But I cannot seem to get it to work.

I keep seeing this:
https://localhost.me.no:7168/api/UnauthorizedApi/Redirecturl#error=interaction_required&error_description=Unfortunately%252C+the+user+may+be+not+logged+in+and%252For+additional+interaction+is+required.

I am sending with this.
http://localhost:8080/oauth/v2/authorize?client_id=190313858991194114@integration_service&scope=openid profile email&redirect_uri=https://localhost.me.no:7168/api/UnauthorizedApi/Redirecturl&response_type=id_token token&id_token_hint=190018611782418435&login_hint=brian@me.no&nonce=Test&prompt=none&ui_locales=en

I forgot to mention I am sending with a insomniac rest and using basic authentication as specified
![image](https://user-images.githubusercontent.com/77129/205276564-350ce279-8c2e-4695-99cb-581b2e9151c7.png)

Can you help with an example of how to get this working or perhaps I am doing something that is not supported?
Thanks

**Comment by fforootd** [comment]
Hi There 

Quick question. Do you want to login a "human" user or a machine?

**Comment by bbakkebo** [comment]
It will be a machine
I have an closed-access api which I give access to customers through a normal username and password.
So this is in place, already, so I need to copy this structure in the new api.

...Well, now that I think of it, I guess I would call them non-human client access.  But I need to handle authorization and getting of the token through my api.

**Reply by fforootd** [reply]
OK, ZITADEL know a concept called "service users" which are like "human users" but without an interactive login.

You have two options to use them:

1. You can create a service user with  [public / private key authentication](https://docs.zitadel.com/docs/guides/integrate/serviceusers)
2. Create a service user and assign it a personal access token, also called [PAT](https://docs.zitadel.com/docs/guides/integrate/pat)

**Reply by bbakkebo** [reply]
Hi and thanks for your quick response.
Ok, I see, this, but if I do this how do I authentication these users?  
I mean which application type must I use to do it?  Is it API private key jwt?

**Reply by fforootd** [reply]
You don't an application typo in ZITADEL for service users.

Just create a service user and create either the jwt private key or a pat.

With the jwt private key the authentication is outlined here https://docs.zitadel.com/docs/guides/integrate/serviceusers#2-create-a-jwt-and-sign-with-private-key

**Reply by bbakkebo** [reply]
Ok, I see. Got it I will check it out then.  Thanks again!

**Reply by bbakkebo** [reply]
Hi again, 

ok, I have got everything working now nicely, but there is something that is not 100% clear with using a PAT.  
Is it possible to renew these tokens from an end-point and the same service user?  
I cannot see how to do this, therefore it seems like the answer is no?

For me, while this would make PAT still useful, but limits them to service to service scenarios behind a firewall, where you do not care so much about expiration.

**Reply by fforootd** [reply]
You can create new keys and remove keys through the [Management API](https://docs.zitadel.com/docs/apis/introduction)

Terraform also should work if you prefer that.

If security is your concern I recommend using the JWT Key approach.

![image](https://user-images.githubusercontent.com/9879976/205636047-f25dab77-77e4-4245-9b4e-6ab12c9829d7.png)

**Comment by bbakkebo** [comment]
Thanks again for your answer. :) 
Yes, I saw this and was thinking I could do it, but the problem is it is a manual process unless I build a gui/api around it, but it would still be manual. 
So the JWT key approach looks the best.  But, I am still trying to get this to work in .net using this:
https://docs.duendesoftware.com/identityserver/v6/tokens/authentication/jwt/ 

Not sure if you know of any solution to this in .net?

**Reply by fforootd** [reply]
If you use .net I recommend you look at https://github.com/smartive/zitadel-net which is maintained by one of our partners.

This also supports the JWT Profile

**Reply by bbakkebo** [reply]
Yes, I am using this and JWT Profile for introspection and it works perfectly.
But the problem is how to generate the private key signed tokens for service users. 
I did not see that they library provides any methods for this?

**Reply by bbakkebo** [reply]
Oh, perhaps I found it now.. I just checked again and found this page https://github.com/smartive/zitadel-net/tree/main/src/Zitadel I think this should be added to your guides as it would have been extremely helpful to see for .net users.

**Reply by bbakkebo** [reply]
Hello again. 

Ok, I went through this and still find it to be lacking and difficult to achieve what I want here.  
I guess the problem is wanting to have the same as I have originally with identity server with basic authentication and issuing JWT's.  So I do not have to disturb my existing users.

So ok, I decided to just use Personal Access Tokens(PAT) as a solution for now.  
But, what I would like to do with using PAT's is set them up so they expire every x number of days and so that my end-user/customer can easily refresh themselves when they expire.
Is this possible to achieve in an automated way?  Ideally it would be great if the customer can send a token request with their old token when their token expires and then receive a new token.  

You can understand that as it stands now with PAT's while I feel that they are secure enough in principle, but if they get intercepted, then I will have a data breach on my api, which is not good.. Of course a thought was put it behind a firewall, but then it is even more painful with having to have customers ip address, etc.
